sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -183,13 +183,14 @@ case class InsertIntoHadoopFsRelationCommand(
         refreshUpdatedPartitions(updatedPartitionPaths)
       }
 
-      // refresh cached files in FileIndex
-      fileIndex.foreach(_.refresh())
-      // refresh data cache if table is cached
-      sparkSession.catalog.refreshByPath(outputPath.toString)
-
       if (catalogTable.nonEmpty) {
+        sparkSession.sessionState.catalog.refreshTable(catalogTable.get.identifier)
```

Comment:
Good catch. new created table's stats is empty, right?

Reply:
yep... so it wont execute this flow... this is what i want to say in my PR https://github.com/apache/spark/pull/22758

Reply:
might be the way i explained was not clear to all

Reply:
This is the reason i asked why in some flow we are initializing the stats and for some flow we are not because of which stats will be none and refreshTable will be never called.
in my PR i told the flow where i saw in insert flow we are not nitializing the stats because of which refreshTable () flow will never be executed.
But before insert command you execute a select statement where stats will be intialized and the relation will be cached, now if you execute insert query refreshTable() will be called as this time the stats will be nonempty

Reply:
May be we should fixed it by: https://github.com/apache/spark/pull/20430

Reply:
Here one more point i want to mention is in hive by default hive.stats.autogather is true, so the DML commands in hive tables will make sure the stats always updated, so i think if we initialize the stats in insert flow , the problems will be solved.

Reply:
if we initialize the stats, then the updateTableStats flow will be executed where we are also updating the table stats and invalidating the cache. and this will ensure the consistency in insert flow.

Reply:
currently our many customer scenarios are getting affected because of this issue.

Reply:
I think this issue shall not be in improvement category, it shall be Critical